### DISCUSSION TITLE
Esp32 web radio, transmit to bluetooth speaker / Compilation errors

### DISCUSSION THREAD

**Discussion by PowerTcX** [pschatzmann/ESP32-A2DP#326] [Q&A] [answered]
Hi, first thank U for a great audio library for esp32.
I'm trying to make an esp32 web radio player, so esp32 transmit url audio to a bluetooth speaker.
I Tried make it from examples, but always got compilation errors.
**_I found that WiFi and Bluetooth coexistence is problematic, so just play .mp3 files from SD card will be fine._**
To solve WiFi/Bluetooth coexistence i planning to use external BT like this -
[Bluetooth Module Audio Transmitter](https://www.aliexpress.com/item/1005003224071356.html?spm=a2g0o.productlist.0.0.315e5e92HLugXp&algo_pvid=51d6fbad-69fc-4eea-b7a2-a88bd47c0a7c&algo_exp_id=51d6fbad-69fc-4eea-b7a2-a88bd47c0a7c-2&pdp_ext_f=%7B%22sku_id%22%3A%2212000024736963568%22%7D&pdp_npi=2%40dis%21EUR%215.19%213.07%21%21%211.69%21%21%402101d68d16694691870712988e54ad%2112000024736963568%21sea&curPageLogUid=67BToFcWydhG)

```
#define USE_HELIX 
#define USE_A2DP

#include "AudioTools.h"
#include "BluetoothA2DPSource.h"
#include "AudioCodecs/CodecMP3Helix.h"

using namespace audio_tools;
BluetoothA2DPSource a2dp_source;  

const char *urls[] = {
  "http://centralcharts.ice.infomaniak.ch/centralcharts-128.mp3",
  "http://centraljazz.ice.infomaniak.ch/centraljazz-128.mp3",
  "http://centralrock.ice.infomaniak.ch/centralrock-128.mp3",
  "http://centralcountry.ice.infomaniak.ch/centralcountry-128.mp3",
  "http://centralfunk.ice.infomaniak.ch/centralfunk-128.mp3"
};
const char *wifi = "Wifi";
const char *password = "PassWord";

URLStream urlStream(wifi, password);
AudioSourceURL source(urlStream, urls, "audio/mp3");
//A2DPStream out = A2DPStream::instance();  // A2DP output - A2DPStream is a singleton!
MP3DecoderHelix decoder;
AudioPlayer player(source, AudioSourceURL, decoder);


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup output - We send the output via A2DP to a "LEXON MINO L" Bluetooth Speaker
  a2dp_source.start("ATH-M50xBT", AudioPlayer);  
  a2dp_source.set_volume(30);

}

void loop() {
  player.copy();
}


```

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/ESP32-A2DP/wiki/WIFI-and-A2DP-Coexistence

But I think you should not get any compilation errors. Are you sure that you have installed all necessary libraries ?

**Reply by PowerTcX** [reply]
All necessary libraries in correct place.

**Comment by PowerTcX** [comment]
Still get a errors -
C:\Users\power\OneDrive\Рабочий стол\Square Studio Export\Landscape\ESP32_DEv_Bluetooth_mp3_stream\ESP32_DEv_Bluetooth_mp3_stream.ino:13:1: error: 'AudioSourceSdFat' does not name a type
C:\Users\power\OneDrive\Рабочий стол\Square Studio Export\Landscape\ESP32_DEv_Bluetooth_mp3_stream\ESP32_DEv_Bluetooth_mp3_stream.ino:14:30: error: 'instance' is not a member of 'audio_tools::A2DPStream'
C:\Users\power\OneDrive\Рабочий стол\Square Studio Export\Landscape\ESP32_DEv_Bluetooth_mp3_stream\ESP32_DEv_Bluetooth_mp3_stream.ino:16:20: error: 'source' was not declared in this scope
C:\Users\power\OneDrive\Рабочий стол\Square Studio Export\Landscape\ESP32_DEv_Bluetooth_mp3_stream\ESP32_DEv_Bluetooth_mp3_stream.ino:16:20: note: suggested alternative: 'fpurge'

exit status 1

Compilation error: 'AudioSourceSdFat' does not name a type

```
#define USE_HELIX 
#define USE_A2DP
#define USE_SDFAT

#include "AudioTools.h"
#include "BluetoothA2DPSource.h"
#include "AudioCodecs/CodecMP3Helix.h"

using namespace audio_tools;  

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSdFat source(startFilePath, ext);
A2DPStream out = A2DPStream::instance();  // A2DP output - A2DPStream is a singleton!
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup output - We send the output via A2DP to a "LEXON MINO L" Bluetooth Speaker
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "JBL GO Essential";
  out.begin(cfg);

  // setup player
  player.setVolume(0.1);
  player.begin();

}

void loop() {
  player.copy();
}
```

**Comment by pschatzmann** [comment]
Where did you get this example from ? It uses old style obsolete syntax:
- remove the USE statements
- add the missing include files

Please use the examples from the examples directory only...
The audio sources can be found in AudioLibs

**Reply by PowerTcX** [reply]
Example from - 
https://www.pschatzmann.ch/home/2021/11/02/streaming-mp3-files-to-a-bluetooth-speaker/

**Comment by pschatzmann** [comment, marked as answer]
Blogs are using the syntax that was valid when the blog was written. 
Please use the updated example from Github

**Reply by PowerTcX** [reply]
Thank U, i found a correct example.
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-a2dp/player-sdfat-a2dp.ino

**Reply by PowerTcX** [reply]
I converted mp3 music to 128kbps, music is playing but sound is not continous, like - 0.5s play, 0.5s silence.
